When using c-ares, resolve ip literals and Windows localhost on our own

apolcyn opened the pull request:
Fixes https://github.com/grpc/grpc/issues/15158

Tested core end2end tests (which use "localhost" as a target) under `GRPC_DNS_RESOLVER=ares`, on a Windows machine having `C:\Windows\System32\drivers\etc\hosts` at it's default setting (with "localhost" commented out).

------

Note updates to this PR noted in:
  * https://github.com/grpc/grpc/pull/16420#issuecomment-415222352
    - Not only manually resolve "localhost" on windows manually, but also resolver ipv4 and ipv6 literals manually: if we don't resolve ip literals manually and attempt to resolve them via c-ares, then we end up making A record queries for raw ipv6 address, AAAA record queries for raw ipv4 addresses, and TXT/SRV record queries for both raw ipv6 and raw ipv4 addresses.
  * https://github.com/grpc/grpc/pull/16420#issuecomment-418932777
    - Skip a couple of unit tests with the c-ares resolver; the c-ares resolver is especially vulnerable to external DNS server query flakes because it doesn't take advantage of the same system caches that the system resolver uses.

PTAL

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/16420)
<!-- Reviewable:end -->

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +208 [None]                                                                           +1.38Ki  +0.0%
  +3.2%    +176 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +176  +3.2%
      [NEW] +1.42Ki grpc_dns_lookup_continue_after_check_localhost_ares_locked                       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.0%    +400 TOTAL                                                                            +1.57Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,970      Total (=)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,051      Total (>)     10,665,042

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +208 [None]                                                                           +1.38Ki  +0.0%
  +3.2%    +176 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +176  +3.2%
      [NEW] +1.42Ki grpc_dns_lookup_continue_after_check_localhost_ares_locked                       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.0%    +400 TOTAL                                                                            +1.57Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,970      Total (=)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,049      Total (>)     10,665,044

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +208 [None]                                                                           +1.38Ki  +0.0%
  +3.2%    +176 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +176  +3.2%
      [NEW] +1.42Ki grpc_dns_lookup_continue_after_check_localhost_ares_locked                       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.0%    +400 TOTAL                                                                            +1.57Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +208 [None]                                                                           +1.38Ki  +0.0%
  +3.2%    +176 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +176  +3.2%
      [NEW] +1.42Ki grpc_dns_lookup_continue_after_check_localhost_ares_locked                       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.0%    +400 TOTAL                                                                            +1.57Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,970      Total (=)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,043      Total (<)     10,665,045

 No significant differences in binary sizes


```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.0%    +208 [None]                                                                           +1.38Ki  +0.0%
  +3.2%    +176 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +176  +3.2%
      [NEW] +1.42Ki grpc_dns_lookup_continue_after_check_localhost_ares_locked                       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.0%    +400 TOTAL                                                                            +1.57Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,970      Total (=)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,044      Total (>)     10,665,042

 No significant differences in binary sizes


```

apolcyn commented:
I've put a do not merge label on this since I'm taking a look at mac and windows failures on the new test. I <i>think</i> that the rest of the source code changes are still ready for a look though, and the problem is in the test or something pre-existing.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +608 [None]                                                                           +2.79Ki  +0.0%
      +0.1%    +576 [Unmapped]                                                                       +2.79Ki  +0.0%
      +3.1%     +32 [None]                                                                                 0  [ = ]
   +14%    +736 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +736   +14%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
      +2.7%     +24 on_hostbyname_done_locked                                                            +24  +2.7%
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.1% +1.33Ki TOTAL                                                                            +3.52Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,584      Total (=)      1,952,584

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,725      Total (<)     10,665,726

 No significant differences in binary sizes


```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,952,584      Total (=)      1,952,584

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,731      Total (>)     10,665,727

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
I've removed the "do not merge" since I this should be ready for a look again.

I updated this PR to also add a `resolve_as_ip_literal` "short circuit" at the top of the c-ares resolution code. This is needed to avoid doing A-record lookups for ipv6 literals and AAAA-record lookups for ipv4 literals. Such lookups are problematic because they add latency and appear to be a source of potential flakiness in tests; originally, the newly enabled tests ran into a couple of timeouts on mac and windows kokoro, and when reproducing these errors, they <i>appear</i> to be due to occasionally not receiving a timely response from the external DNS server.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +480 [None]                                                                           +7.14Ki  +0.1%
   +14%    +752 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +752   +14%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.1% +1.22Ki TOTAL                                                                            +7.89Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,968      Total (=)      1,955,968

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,884      Total (>)     10,843,881

 No significant differences in binary sizes


```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +480 [None]                                                                           +7.14Ki  +0.1%
   +14%    +752 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +752   +14%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.1% +1.22Ki TOTAL                                                                            +7.89Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,968      Total (=)      1,955,968

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,882      Total (<)     10,843,883

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
PTAL, this should be ready for a look

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +480 [None]                                                                           +7.16Ki  +0.1%
   +14%    +752 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +752   +14%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
      [NEW]    +110 strhtons                                                                            +110  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 maybe_resolve_localhost_manually_locked                                               +3  [NEW]

  +0.1% +1.22Ki TOTAL                                                                            +7.91Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,955,968      Total (=)      1,955,968

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,843,877      Total (<)     10,843,881

 No significant differences in binary sizes


```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +573 [None]                                                                           +3.17Ki  +0.0%
      +0.1%    +541 [Unmapped]                                                                       +3.17Ki  +0.0%
      +3.1%     +32 [None]                                                                                 0  [ = ]
   +12%    +640 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +640   +12%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
  +6.4%    +115 src/core/ext/filters/client_channel/parse_address.cc                                +115  +6.4%
      [NEW]    +110 grpc_strhtons                                                                       +110  [NEW]
       +13%      +5 [Unmapped]                                                                            +5   +13%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 grpc_ares_maybe_resolve_localhost_manually_locked                                     +3  [NEW]

  +0.1% +1.31Ki TOTAL                                                                            +3.92Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,960,882      Total (=)      1,960,882

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,881,734      Total (>)     10,881,530

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +573 [None]                                                                           +3.18Ki  +0.0%
      +0.1%    +541 [Unmapped]                                                                       +3.18Ki  +0.0%
      +3.1%     +32 [None]                                                                                 0  [ = ]
   +12%    +640 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +640   +12%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
  +6.4%    +115 src/core/ext/filters/client_channel/parse_address.cc                                +115  +6.4%
      [NEW]    +110 grpc_strhtons                                                                       +110  [NEW]
       +13%      +5 [Unmapped]                                                                            +5   +13%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 grpc_ares_maybe_resolve_localhost_manually_locked                                     +3  [NEW]

  +0.1% +1.31Ki TOTAL                                                                            +3.93Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,960,882      Total (=)      1,960,882

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,881,738      Total (>)     10,881,530

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
Forced update with an `s/strhtons/grpc_strhtons/` update within `grpc_ares_wrapper_windows.cc`

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                        FILE SIZE
 ++++++++++++++ GROWING                                                                          ++++++++++++++
  +0.1%    +573 [None]                                                                           +3.17Ki  +0.0%
      +0.1%    +541 [Unmapped]                                                                       +3.17Ki  +0.0%
      +3.1%     +32 [None]                                                                                 0  [ = ]
   +12%    +640 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc        +640   +12%
      [NEW] +1.42Ki grpc_dns_lookup_ares_continue_after_check_localhost_and_ip_literals_locked       +1.42Ki  [NEW]
       +27%      +9 grpc_cancel_ares_request                                                              +9   +27%
  +6.4%    +115 src/core/ext/filters/client_channel/parse_address.cc                                +115  +6.4%
      [NEW]    +110 grpc_strhtons                                                                       +110  [NEW]
       +13%      +5 [Unmapped]                                                                            +5   +13%
   +84%     +16 src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper_posix.     +16   +84%
      [NEW]     +13 [Unmapped]                                                                           +13  [NEW]
      [NEW]      +3 grpc_ares_maybe_resolve_localhost_manually_locked                                     +3  [NEW]

  +0.1% +1.31Ki TOTAL                                                                            +3.92Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,960,906      Total (=)      1,960,906

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,881,738      Total (>)     10,881,534

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

apolcyn commented:
Forced update to `#include "src/core/ext/filters/client_channel/parse_address.h"` in `grpc_ares_wrapper_windows.cc`

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,960,906      Total (=)      1,960,906

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,881,731      Total (>)     10,881,530

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```
